### ISSUE TITLE
Enum implementing Equatable related classes

### ISSUE THREAD

**Issue by FMorschel** [felangel/equatable#147]
I've created a class that only takes `Enum`s as parameters. I figured I could create a third `Enum` where I would manually put every option so they have a better naming.

E.g.:

### Configuration
```dart
enum A {
  a,
  b;
}
enum B {
  c,
  d;
}

class Class with EquatableMixin {
  const EveryDayOfYear({required this.aValue, required this.bValue});

  final A aValue;
  final B bValue;
  
  @override
  List<Object?> get props => [aValue, bValue];
}

enum C implements Class {
  ac(Class(aValue: A.a, bValue: B.c)),
  ad(Class(aValue: A.a, bValue: B.d)),
  bc(Class(aValue: A.b, bValue: B.c)),
  bd(Class(aValue: A.b, bValue: B.d));

  const C(this._handler);

  final Class _handler;

  @override
  A get aValue => _handler.aValue;

  @override
  B get bValue => _handler.bValue;

  @override
  List<Object?> get props => [aValue, bValue];
}
```

### Objective

```dart
final instance = Class(aValue: A.a, bValue: B.c);
instance == C.ac; // I would like something so this operation returns true.
```

### Context

As I said [here](https://stackoverflow.com/a/73407104/9576870):

 > As commented by [@Levi-Lesches](https://github.com/Levi-Lesches) [here](https://github.com/dart-lang/language/issues/2414#issuecomment-1219732598) the answer to my problem was to override my `operator ==`:
 > 
 > ```dart
 >   @override
 >   // ignore: hash_and_equals, already implemented by EquatableMixin
 >   bool operator ==(Object other) {
 >     return (super == other) ||
 >         ((other is Class) &&
 >             (aValue == other.aValue) &&
 >             (bValue == other.bValue));
 >   }
 > ```
This solved my problem and is fair since my class instance is not an `Enum`, but my enum constant is actually my class instance.

### Suggestion

Enable some kind of option or by default so that extended classes are also equal to the base class not by testing the `runtimeType` but by testing if the object `is` from the current class/equatable related.

**Comment by FMorschel** [comment]
This would also solve #100

**Comment by FMorschel** [comment]
I've tested:

<details>
  <summary>Current case</summary>

  All the tests on this code pass
  ```dart
  import 'package:equatable/equatable.dart';
  import 'package:test/test.dart';
  
  class A extends Equatable {
    A(this.a);
  
    final int a;
  
    @override
    List<Object?> get props => [a];
  }
  
  class B with EquatableMixin {
    B(this.b);
  
    final int b;
  
    @override
    List<Object?> get props => [b];
  }
  
  void main() {
    test('Testing A', () {
      final a1 = A(1);
      final a2 = A(1);
      final a3 = A(2);
      expect(a1, equals(a2));
      expect(a1, isNot(a3));
      expect(a2, isNot(a3));
    });
    test('Testing B', () {
      final b1 = B(1);
      final b2 = B(1);
      final b3 = B(2);
      expect(b1, equals(b2));
      expect(b1, isNot(b3));
      expect(b2, isNot(b3));
    });
  }
  ```

  ### WARNING
  But, these do as well:
  ```dart
  //...
  class A2 extends Equatable {
    A2(this.a);

    final int a;

    @override
    Type get runtimeType => A;

    @override
    List<Object?> get props => [a];
  }
  //...
  void main() {
    //...
    test('Testing A2', () {
      final a1 = A2(1);
      final a2 = A2(2);
      final a = A(1);
      expect(a1, equals(a));
      expect(a1, isNot(a2));
      expect(a2, isNot(a));
    });
    //...
  }
  ```

  ### WARNING 2
  But these all fail:
  ```dart
  //...
  enum E implements B {
    e1,
    e2;
  
    @override
    bool? get stringify => null;
  
    @override
    int get b => index;
  
    @override
    List<Object?> get props => [b];
  }
  //...
  void main() {
    //...
    test('Testing E', () {
      final e1 = E.e1;
      final e2 = E.e2;
      final b1 = B(0);
      expect(e1, equals(b1));
      expect(b1, isNot(e1));
      expect(b1, isNot(e2));
    });
    //...
  }
  ```


</details>
<details>
  <summary>My suggested case</summary>

  When the `==` operator is as I've configured below:
  Mixin:
  ```dart
  mixin EquatableMixin<T extends Object> on Object {
    //...
    @override
    bool operator ==(Object other) {
      return identical(this, other) ||
          other is EquatableMixin<T> && //CHANGED this line so the generic is used and removed the `runtimeType`
              equals(props, other.props);
    }
    //...
  }
  ```
  Equatable:
  ```dart
  abstract class Equatable<T extends Object> {
    //...
    @override
    bool operator ==(Object other) =>
        identical(this, other) ||
        other is Equatable<T> && //CHANGED this line so the generic is used and removed the `runtimeType`
            equals(props, other.props);
    //...
  }
  ```
  All the tests on this code pass
  ```dart
  import 'package:equatable/equatable.dart';
  import 'package:test/test.dart';
  
  class A extends Equatable<A> {
    A(this.a);
  
    final int a;
  
    @override
    List<Object?> get props => [a];
  }
  
  class B with EquatableMixin<B> {
    B(this.b);
  
    final int b;
  
    @override
    List<Object?> get props => [b];
  }
  
  enum E implements B {
    e1,
    e2;
  
    @override
    bool? get stringify => null;
  
    @override
    int get b => index;
  
    @override
    List<Object?> get props => [b];
  }
  
  void main() {
    test('Testing A', () {
      final a1 = A(1);
      final a2 = A(1);
      final a3 = A(2);
      expect(a1, equals(a2));
      expect(a1, isNot(a3));
      expect(a2, isNot(a3));
    });
    test('Testing B', () {
      final b1 = B(1);
      final b2 = B(1);
      final b3 = B(2);
      expect(b1, equals(b2));
      expect(b1, isNot(b3));
      expect(b2, isNot(b3));
    });
    test('Testing E', () {
      final e1 = E.e1;
      final e2 = E.e2;
      final b1 = B(0);
      expect(e1, equals(b1));
      expect(b1, isNot(e1));
      expect(b1, isNot(e2));
    });
  }
  ```

  ### WARNING
  But, these do as well:
  ```dart
  //...
  class A2 extends Equatable<A> {
    A2(this.a);

    final int a;

    @override
    List<Object?> get props => [a];
  }
  //...
  void main() {
    //...
    test('Testing A2', () {
      final a1 = A2(1);
      final a2 = A2(2);
      final a = A(1);
      expect(a1, equals(a));
      expect(a1, isNot(a2));
      expect(a2, isNot(a));
    });
    //...
  }
  ```
</details>

There are some possible errors in the current case (problem I presented on this issue and #100). I know it's not something everyone will do, overriding the getter of the `runtimeType` to any other type that has no real relation, but there is still a chance for this to happen and the equality to fail.

My current suggestion would be a breaking change if done this way, but creating a new class and a new mixin for this, could be a solution to that. And this would give the control for the equality for the user, similar to when using `Comparable` or some other class that uses generics. And this would keep the subclasses all equal to their base class but the other way around (like on `enum`s cases) sometimes not.

**Comment by FMorschel** [comment]
The only thing I was thinking here was for my suggestions to work the `props` should have [mustCallSuper](https://api.flutter.dev/flutter/meta/mustCallSuper-constant.html) and this should be explained to be written like:
```dart
@override
List<Object?> get props => [...super.props, newProps];
```

**Comment by escamoteur** [comment]
not sure if there is a real difference of using the runtime type or using `is` because `is` will have to check the runtime type of the `other` object. Also I think we are no longer talking of real equality then anymore if an instance of a derived class with the same property values is seen as equal.

**Comment by felangel** [comment]
This should be resolved by #175 since you'll no longer need to manually specify props. You can already try it out on the master channel using the #174 branch (still wip).
